From aad5995721823e55e381c1baa9ce3b5360ad2c3b Mon Sep 17 00:00:00 2001
From: Petr Machata <petrm@mellanox.com>
Date: Wed, 20 Nov 2019 17:01:19 +0200
Subject: [PATCH 1/3] mlnx_qos: Support default priority assignment

When APP TLV selector 1 (EtherType) is used with PID of 0, the
corresponding entry specifies "default application priority [...] when
application priority is not otherwise specified."

Introduce "--defltprio" command to allow setting and deleting these APP
entries.

Signed-off-by: Petr Machata <petrm@mellanox.com>
---
 ofed_scripts/utils/dcbnetlink.py |  7 +++++++
 ofed_scripts/utils/mlnx_qos      | 32 ++++++++++++++++++++++++++++++++
 2 files changed, 39 insertions(+)

diff --git a/ofed_scripts/utils/dcbnetlink.py b/ofed_scripts/utils/dcbnetlink.py
index 220709f..3fc44ce 100644
--- a/ofed_scripts/utils/dcbnetlink.py
+++ b/ofed_scripts/utils/dcbnetlink.py
@@ -109,6 +109,8 @@ DCB_ATTR_DCB_BUFFER = 10
 DCB_ATTR_IEEE_APP_UNSPEC = 0
 DCB_ATTR_IEEE_APP = 1
 
+IEEE_8021QAZ_APP_SEL_ETHERTYPE	= 1
+
 class DcbnlHdr:
 	def __init__(self, len, type):
 		self.len = len
@@ -504,6 +506,11 @@ class DcbAppTable:
 			if temp != "":
 				print(temp)
 
+	def printDefaultPriority(self):
+		for app in self.apps.values():
+			if app.selector == IEEE_8021QAZ_APP_SEL_ETHERTYPE:
+				print("\tprio:%d" % app.priority)
+
 	def delAppEntry(self, ctrl, selector):
 		for i in range(len(self.apps)):
 			if self.apps[i].selector == selector:
diff --git a/ofed_scripts/utils/mlnx_qos b/ofed_scripts/utils/mlnx_qos
index 3d134cd..d545251 100755
--- a/ofed_scripts/utils/mlnx_qos
+++ b/ofed_scripts/utils/mlnx_qos
@@ -100,6 +100,9 @@ parser.add_option("--trust", dest="trust",
 parser.add_option("--dscp2prio", dest="dscp2prio",
                   help="set/del a (dscp,prio) mapping. Example 'set,30,2' maps dscp 30 to priority 2. " +
                        "'del,30,2' deletes the mapping and resets the dscp 30 mapping back to the default setting priority 0.")
+parser.add_option("--defltprio", dest="defltprio",
+                  help="set/del default priority. Example 'set,2' sets default priority to 2. " +
+                       "'del,2' deletes this setting.")
 parser.add_option("--cable_len", dest="cable_len",
                   help="set cable_len for buffer's xoff and xon thresholds")
 parser.add_option("--prio2buffer", dest="prio2buffer",
@@ -293,6 +296,8 @@ def pretty_print(prio_tc, tsa, tcbw, ratelimit, pfc_en, trust, pfc_delay, buffer
 			print_dscp2prio_sysfs()
 		else:
 			appTable.printAppSelector(IEEE_8021QAZ_APP_SEL_DSCP)
+	print("default priority:")
+	appTable.printDefaultPriority()
 
 	if (not buffer_error):
 		msg = "Receive buffer size (bytes): "
@@ -462,6 +467,33 @@ except:
 		if (not is_trust_sysfs):
 			appTable = DcbAppTable()
 
+# ********* defltprio command ******************************
+try:
+	if options.defltprio:
+		action, prio = options.defltprio.split(",")
+		prio = int(prio)
+
+		if action not in ("set", "del") or prio < 0 or prio > 7:
+			raise Exception()
+except:
+	print("Invalid defltprio command. Refer to the help.")
+	sys.exit(1)
+
+try:
+	if options.defltprio:
+		if action == "set":
+			ctrl.set_ieee_app(IEEE_8021QAZ_APP_SEL_ETHERTYPE,prio,0)
+		elif action == "del":
+			ctrl.del_ieee_app(IEEE_8021QAZ_APP_SEL_ETHERTYPE,prio,0)
+
+	appTable = ctrl.get_ieee_app_table()
+except:
+	if options.defltprio:
+		print("defltprio command failed")
+		sys.exit(1)
+	else:
+		appTable = DcbAppTable()
+
 # ********* trust command ******************************
 if options.trust:
 	if (options.trust != "dscp") and (options.trust != "pcp"):
-- 
2.17.1

